ext/storage_sources/azure_store/azure_connection.cpp

```diff
@@ -39,6 +39,13 @@ azure_connection::azure_connection(const std::string &bucket_name, const std::st
         std::getenv("AZURE_STORAGE_CONNECTION_STRING"), bucket_name)),
       _bucket_name(bucket_name), _object_prefix(obj_prefix)
 {
+    // Confirm that we can access the bucket, else fail.
+    bool exists;
+    int ret = bucket_exists(exists);
+    if (!exists)
+        std::cerr << _bucket_name + " : No such bucket." << std::endl;
+    if (ret != 0)
+        std::cerr << _bucket_name + " : Unable to access bucket." << std::endl;
 }
```

Comment:
I think there's a better approach here. Thinking about what happens either if the bucket doesn't exist or something else goes wrong, the constructor will work fine but leave the object in an invalid state. You're correct that you can't return a value here, the most appropriate tool here is to throw an exception.

We can talk about when that's appropriate - for example, it might be valid to have an `AzureConnection` where the bucket does not exist, I'm not sure what your needs are here.

Reply:
thank you for the suggestion. Implemented :)

Reply:
No problems! One last nitpick with this, these should be a `std::runtime_error` - an `std::invalid_argument` is a logic error, which implies the caller has violated some logical precondition and it should be preventable.

(You can take a look at the docs to see examples of the kind of issue that `invalid_argument` is generally used for: https://en.cppreference.com/w/cpp/error/invalid_argument)

Reply:
I considered that it should throw an `invalid_argument` because the user could:
- give a bucket name that already exists and so it won't create that bucket
```
if (exists == false)
        throw std::invalid_argument(_bucket_name + " : No such bucket.");
``` 

But also, a `runtime_error` would be more suitable in the case that a user:
- give a bucket name of a recently deleted container - context: the deleted bucket is marked for deletion and only officially deleted when garbage collection comes around to clear it, which the azure docs indicate can sometimes take several minutes.
```
if (ret != 0)
        throw std::runtime_error(_bucket_name + " : Unable to access bucket.");
```

I also took into consideration that s3 implementation throw an `invalid_argument` for both cases. Can I throw an `invalid_argument` in the case `exists == false` and a `runtime_error` in the case `ret != 0`

Quick side note:
I've moved the `ret != 0` condition above `exists == false` because with the way I've implemented `bucket_exists`, it would always throw `exists == false` and never throw `ret != 0`

Reply:
resolved